### ISSUE TITLE
Sharing driver between notebooks

### ISSUE THREAD

**Issue by idanz** [andypetrella/spark-notebook#277]
Currently it seems the server is spawning a new driver for each notebook.
While this is good for isolation, it causes issues of memory allocation, as driver often need at least 8GB to support operations like BroadcastJoin and just collecting data back.
This means that it's hard to scale the usage of the spark notebook.

The Zeppelin project, on the other hand, uses a single driver for all notebooks, thus solving this issue.
Would it be possible to support driver sharing for the spark notebook as well?

**Comment by andypetrella** [comment]
Thanks @idanz to share this use case.

Actually, the separation of processes is one thing important to be able to connect to several cluster using the same server of course.
On the other hand, you're right, a spark context might be eager in memory, specially in local or with joins.

So sharing sparkContext, as you said, would allow a better management of the local memory, but having only one prevent a server to connect to several clusters. We might need an intermediate level where we could share some sparkContext → which would be cool.

For now a sparkContext is part of a REPL started by the server in a process, the REPL have a session associated to it -- for web socket and so forth, but specially to be able to reconnect if you close the tab, or if you want to connect from elsewhere.

We can think in a way that we can start a REPL/SparkContext, and when we open a notebook we could ask it to connect to the same session rather starting a new one.

Although I think it makes sense (for instance it would speedup the starting time!), it doesn't really solve the scalability right? I mean, nothing really prevents one to use eagerly the driver's memory → imagine three notebooks doing a BroadcaseJoin with diff datasets (looks like the HPC age ^^).

That's why I have this idea for a while, thus I'll create the ticket after, to have the process not started locally only, but having them started via a cluster manager like mesos (marathon, aurora maybe), yarn (not sure how, not really done for that :-S), Borg. 
Then it's a matter of the notebook/SparkContext (if we do what I proposed above) to ask for resources.

WDYT?

**Comment by idanz** [comment]
Not sure I completely follow, are you talking about running the driver on the cluster as well?
That would be cool and more scalable indeed.
Personally I use the yarn manager as most spark users AFAIK, so running in yarn-cluster mode would solve it.
However, comparing with Zeppelin and the spark thrift server, both of these solutions run on single shared driver. Personally I prefer your project over Zeppelin but this issue is a blocker for me.
So personally I would selfishly prefer what will be supported sooner..

**Comment by andypetrella** [comment]
Yes this is what I meant :-D. And yes, yarn-master will work, but the driver's memory (the guy that dies with `collect`) will be still locally until we can spawn the process on a node, however, I don't think yarn supports that right?

Anyway, in order to have it locally, this share I mean, we'll need to do some work, but we can start it and I can include it into the roadmap (that we'll do soon I hop) for 0.7.0.

Do you know if you or some of your guyz in your company would be likely available to help/support/test :-P? (yes, trying to recruit ^^). 

cheers

**Comment by andypetrella** [comment]
oh and @idanz thanks for the kind words about your prefs, very appreciated!

**Comment by idanz** [comment]
there is a yarn-cluster mode, and we actually use that, but not with the REPL console (only with spark submit). For that we would need to let the driver run somewhere and then find it on the cluster and talk to it by rpc when submitting paragraphs (or even doing code completion).

I'll pass it around my company, maybe someone will want to join in.
Thanks!

**Comment by andypetrella** [comment]
Actually, the notebook server does that code sending, so it'll need to know where the process is deployed (the host and the akka remote port) and it'll run fine. But some works still needs to be done.

Let's first focus on how to allow some notebooks to share a sparkContext. Actually, this a very good example of integration with the spark job server I guess → maybe @velvia can give it a think? ^^

**Comment by velvia** [comment]
I'll have to understand more about the notebook first  :)      especially how the code is sent, how results are sent back etc.    The way sharing in job server works is that a context is created with a POST request, then individual "jobs" can be run against the shared context.  One thing to think about is how RDDs are shared.  In the job server, there is a multi-thread-safe facility to name and register RDDs, which is designed to make it easy for different "jobs" to operate on named RDDs.
